### PR TITLE
[RLlib] Document and extend action mask example.

### PR DESCRIPTION
**Pull request by stefanbschneider**
## Why are these changes needed?

Adds comments to document the action masking example.
Also extends the example to allow training manually without Ray tune.
And adding an option to disable masking, illustrating that invalid actions are indeed selection without masking.

## Related issue number

Ongoing effort of improving and extending all examples.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(
   - [x] Local, manual tests

### CODE REVIEW

File: rllib/examples/action_masking.py [hunk 1]
```diff
@@ -1,61 +1,121 @@
+"""
+Example using action masking to select/avoid actions depending on the current
+observation. This is useful in many practical scenarios, where different
+actions are available in different time steps.
+Blog post explaining action masking: https://boring-guy.sh/posts/masking-rl/
```
Comment (richardliaw) [inline rllib/examples/action_masking.py:5]:
```suggestion
"""Example showing how to use "action masking" in RLlib.

"Action masking" allows the agent to select actions based on the current
observation. This is useful in many practical scenarios, where different
actions are available in different time steps.
Blog post explaining action masking: https://boring-guy.sh/posts/masking-rl/
```

File: rllib/examples/action_masking.py [hunk 2]
```diff
@@ -1,61 +1,121 @@
...
+
+RLlib supports action masking, i.e., disallowing these actions based on the
+observation, by slightly adjusting the environment and the model as shown in
+this example.
+Here, the ActionMaskEnv wraps an underlying environment (here, RandomEnv),
+defining only a subset of all actions as valid based on the environment's
+observations. If an invalid action is selected, the environment raises an error
+- this must not happen!
+The environment constructs Dict observations, where obs["observations"] holds
+the original observations and obs["action_mask"] holds the valid actions.
```
Comment (richardliaw) [inline rllib/examples/action_masking.py:15]:
```suggestion
RLlib supports action masking, i.e., disallowing these actions based on the
observation, by slightly adjusting the environment and the model as shown in
this example.

Here, the ActionMaskEnv wraps an underlying environment (here, RandomEnv),
defining only a subset of all actions as valid based on the environment's
observations. If an invalid action is selected, the environment raises an error
- this must not happen!

The environment constructs Dict observations, where obs["observations"] holds
the original observations and obs["action_mask"] holds the valid actions.
```

File: rllib/examples/action_masking.py [hunk 3]
```diff
@@ -1,61 +1,121 @@
...
+them. This only works with discrete actions.
+
+---
+Run this example with defaults (using Tune and action masking):
+$ python action_masking.py
+Then run again without action masking, which will likely lead to errors due to
+invalid actions being selected (ValueError "Invalid action sent to env!"):
+$ python action_masking.py --no-masking
+Other options for running this example:
+$ python action_masking.py --help
```
Comment (richardliaw) [inline rllib/examples/action_masking.py:28]:
```suggestion
Run this example with defaults (using Tune and action masking):

  $ python action_masking.py

Then run again without action masking, which will likely lead to errors due to
invalid actions being selected (ValueError "Invalid action sent to env!"):

  $ python action_masking.py --no-masking

Other options for running this example:

  $ python action_masking.py --help
```

**Comment by richardliaw** [comment]
nice work here ;)

File: rllib/examples/models/action_mask_model.py [hunk 4]
```diff
@@ -35,6 +35,11 @@ def __init__(self, obs_space, action_space, num_outputs, model_config,
             orig_space["observations"], action_space, num_outputs,
             model_config, name + "_internal")
 
+        # disable action masking --> will likely lead to invalid actions
+        self.no_masking = False
+        if "no_masking" in model_config["custom_model_config"]:
+            self.no_masking = model_config["custom_model_config"]["no_masking"]
+
```
Comment (sven1977) [inline rllib/examples/models/action_mask_model.py:42]:
```suggestion
        # disable action masking --> will likely lead to invalid actions
        self.no_masking = model_config["custom_model_config"].get("no_masking", False)

```

File: rllib/examples/action_masking.py [hunk 5]
```diff
@@ -1,61 +1,129 @@
+"""Example showing how to use "action masking" in RLlib.
```
Comment (sven1977) [inline rllib/examples/action_masking.py:1]: Nice!

**Review by sven1977 — APPROVED** [review]
Thanks for this PR, @stefanbschneider ! Really nice.

**Comment by sven1977** [comment]
Waiting for tests, for merging ...

**Pull request merged**